### ISSUE TITLE
Did you figured out how to train style tokens?

### ISSUE THREAD

**Issue by creotiv** [jinhan/tacotron2-gst#5]
Hi,

Im trying to use GST in my tacotron2 project, but problem that no matter how i try, gst tokens not working s intended.
They doesn't add any style to the audio. Maybe there are some tricks not mentioned in the paper?
Can u share your experience and pitfalls during the training? 

Many thanks)

**Comment by jinny1208** [comment]
Hi, I am experiencing the same problem. I have been looking at various repositories, but there does not seem to be a clear solution. Do you think that the problem lies with the dataset that is being used to train the model? That is, do you think you need to use a multispeaker dataset? What kind of dataset are you using and have you fixed the problem?

**Comment by creotiv** [comment]
@jinny1208 I tried many different GST realizations, different datasets with small variation, high variation, one speaker, multi-speaker, with Diagonal guided attention and without. In all cases i was unable to achieve any positive results with Styling.
Also i tried this https://github.com/NVIDIA/mellotron which also has GST, but still got problem with getting it work as expected in the paper.

**Comment by creotiv** [comment]
Also guys here https://habr.com/ru/company/ashmanov_net/blog/528296/ also was saying that they cant make it work

**Comment by jinny1208** [comment]
I really don't understand how people got their repos like https://github.com/KinglittleQ/GST-Tacotron to get GST to work. The repo that I linked, which you may have looked at, is the one that this repo references. Although the structure of the repo is a bit different, the code seems to be the same.

**Comment by jinny1208** [comment]
So basically from the things that you tried, the problem is definitely not the dataset and is something in the code...If I try something out and it works, I will definitely let you know. It might take some time because I may need to retrain my model to reflect the new changes in the code. Please let me know if you are able to get GST to work and what you did to do so.

**Comment by creotiv** [comment]
From my point i think problem in the training process where
embedding vanishing, why i dont know it needs deep research to
understand that, on which i dont have resources.
So I'll rely on luck)

Ok i will let you know if i find something.

Kindly yours,
Andrey Nikishaev

Areas         ML/DS/CV/Soft Dev/BizDev/Growth Hacking/Customer Rel/IT
LinkedIn      http://ua.linkedin.com/in/creotiv
GitHub        http://github.com/creotiv
Slideshare  https://www.slideshare.net/anikishaev/
Skype         creotiv.in.ua
Mobile        +380632410666

**Comment by creotiv** [comment]
@jinny1208 i found that my reference encoder are giving the same vector for different mels, thus styles not trained at all.
Trying to understand why this happening

**Comment by jinny1208** [comment]
@creotiv Using the jupyter notebook in this repo, I did a retest on the "condition on ref_audio" without retraining my model and by using two different reference audios on the same input text. When I took a closer look, the two synthesized audios sounded a bit different from each other (prosody wise) and did in fact, produce slightly different mel spectrograms, as well as tensors. 

It seems to be training somehow, but as you said a couple of days ago, the embedding does seem to vanish with training.

**Comment by creotiv** [comment]
@jinny1208 i found a problem why it was not training. You should train network on one sentence only. Two or more sentences has different prosody, thus it not able to get global style from reference mel, cause there are few of them.

Another thing is if you are using text predicted GST, dont forget to detach input to it and target, so gradient dont go to the main model.

After i fix this i was able to get different styles tooens.

But still i was unable to train network on dataset with high prosody variation, maybe need more tokens and heads on attention of gst.

**Comment by jinny1208** [comment]
>  i found a problem why it was not training. You should train network on one sentence only. Two or more sentences has different prosody, thus it not able to get global style from reference mel, cause there are few of them.

I see.That explains things. Thanks a lot!

> Another thing is if you are using text predicted GST, dont forget to detach input to it and target, so gradient dont go to the main model.

Do you mind elaborating? By detaching input, do you mean in the model.py?

**Comment by creotiv** [comment]
You have two inputs that can backprop gradient.
one goes as an input to TPGST model, another is GST model output, that used
by tpgst as a target. both of them should be detached,
to train only TPGST from it gradient

Kindly yours,
Andrey Nikishaev

Areas         ML/DS/CV/Soft Dev/BizDev/Growth Hacking/Customer Rel/IT
LinkedIn      http://ua.linkedin.com/in/creotiv
GitHub        http://github.com/creotiv
Slideshare  https://www.slideshare.net/anikishaev/
Skype         creotiv.in.ua
Mobile        +380632410666

**Comment by creotiv** [comment]
you can check it here https://github.com/creotiv/RussianTTS-Tacotron2/tree/tpgst2
